### PR TITLE
Snapshot based ensemble learning for KGE

### PR DESCRIPTION
**Pull request by sapkotaruz11**
# Learning Rate Scheduler Overview

The **`LRScheduler`** is a training callback designed to manage **learning rate scheduling**, **model snapshotting**, and **ensemble evaluation** while training KGE models. This approach is based on the paper [SnapE](https://dl.acm.org/doi/10.1007/978-3-031-77844-5_1)

## Key Capabilities

- **Dynamic Learning Rate Scheduling**  
  Supports multiple strategies including Cosine Cyclic Annealing (CCA),  Max-Min Cyclic Cosine Learning Rate (MMCCLR), and their **deferred variants** with warmup phases.

- **Deferred Warmup Support**  
  Deferred schedulers delay learning rate cycling until after an initial warmup period, allowing models to stabilize early in training.

- **Model Snapshotting**  
  Automatically saves model checkpoints at regular intervals, typically aligned with the end of learning rate cycles or distributed across the active phase.

- **Loss-Based Snapshot Weighting**  
  Each snapshot is scored by its training loss, and weights for ensemble averaging are calculated inversely proportional to these losses.

- **Test-Time Ensemble Evaluation**  
  At the end of training, saved snapshots are loaded, optionally weighted, and evaluated as an ensemble for improved performance on the test set.

## Workflow Summary

1. Initialize the scheduler with the selected strategy.
2. Adjust the learning rate at each training batch.
3. Save snapshots at scheduled intervals.
4. Compute loss-based weights for snapshots (if enabled).
5. Evaluate an ensemble of saved models on the test set.

## Output Artifacts

- A directory of saved model snapshots.
- A JSON report with configuration details, snapshot losses, and ensemble evaluation metrics.

## Extension for SWA
- Added the SWA start epoch as a argument, SWA started at later epochs was found to be more effective

### CODE REVIEW

File: dicee/config.py [hunk 1]
```diff
@@ -152,6 +152,13 @@ def __init__(self, **kwargs):
 
         self.n_epochs_eval_model: str = "val_test"
         """Evaluating link prediction performance on data splits while performing periodic evaluation."""
+        
+        self.adaptive_lr = dict()
+        """Adaptive learning rate parameters, e.g., {"lr_decay": 0.1}"""
+
+        self.swa_start_epoch: int = 1
```
Comment (Demirrr) [inline dicee/config.py:159]: Please initialize the swa_start_epoch as None so that the config does not confuses people

File: dicee/scripts/run.py [hunk 2]
```diff
@@ -128,6 +128,18 @@ def get_default_arguments(description=None):
...
                         help='degree for polynomial embeddings')
+    
+    # Learning rate scheduling with configuration
+    parser.add_argument("--adaptive_lr", type=json.loads, default={},
+                        help='Enable adaptive learning rate scheduling with configuration. '
+                             'Example: {"scheduler_name": "cca", "lr_min": 0.01, "num_cycles": 10, '
+                             '"weighted_ensemble": true, "n_snapshots": 5}. '
+                             'Available schedulers: cca, mmcclr, deferred_cca, deferred_mmcclr')
+    
+    parser.add_argument("--swa_start_epoch", type=int, default=1,
```
Comment (Demirrr) [inline dicee/scripts/run.py:139]: Please set the default value None and remove additinal new lines before

**Review by Demirrr — CHANGES_REQUESTED** [review]
Thank you for the PR. It looks great. There are only two minor changes required :)

Could you please add unit and regression tests?

**Review by Demirrr — APPROVED** [review]
Thank you for the changes!

**Comment by sapkotaruz11** [comment]
> Thank you for the changes! Looking good. You might want to add few tests so that none will break your input :)

 @Demirrr   Yes, I will definitely add some tests before merging to the develop. Thank you for the review.

**Comment by sapkotaruz11** [comment]
@Demirrr it seems TorchCPUTrainer does not support SWA callback and also some of the training hooks as in PL trainer; however there seems to be no check for that? Is that actually the case?

**Comment by sapkotaruz11** [comment]
The [line](https://github.com/dice-group/dice-embeddings/blob/b31c18f7471495d0a363271cf1d71a96c39f765f/tests/test_adaptive_swa.py#L35) shows that SWA is actually commented out.

**Comment by Demirrr** [comment]
You are absolutely right! Thank you for pointing out. Could you please investigate this error if you could find time?

**Comment by sapkotaruz11** [comment]
@Demirrr There are some minor changes to the files after adding the regression tests, should the new changes be reviewed?
